### PR TITLE
fix(npm): resolve registry URL using `packageName`, if set

### PR DESCRIPTION
**Pull request by haines**
## Changes

This PR uses a dependency's `packageName` (if set) in preference to `depName` when resolving the registry URL from Yarn config.

## Context

Fixes a bug (as described in https://github.com/renovatebot/renovate/discussions/28245#discussioncomment-9020088) where Renovate does not respect registry settings for the `@yarnpkg` scope when attempting to update `yarn`, because the dependency name (`yarn`) is different to the package name (`@yarnpkg/cli`).

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/extract/index.spec.ts [hunk 1]
```diff
@@ -293,6 +293,28 @@ describe('modules/manager/npm/extract/index', () => {
       ).toBeArrayIncludingOnly(['https://registry.example.com']);
     });
 
+    it('resolves registry URLs using the package name if set', async () => {
+      fs.readLocalFile.mockImplementation((fileName): Promise<any> => {
+        if (fileName === '.yarnrc.yml') {
+          return Promise.resolve(
+            'npmScopes:\n  yarnpkg:\n    npmRegistryServer: https://registry.example.com',
+          );
```
Comment (secustor) [inline lib/modules/manager/npm/extract/index.spec.ts:301]:
```suggestion
          return Promise.resolve(codeblock`
          npmScopes:
            yarnpkg:
                npmRegistryServer: https://registry.example.com
          `);
```

File: lib/modules/manager/npm/extract/index.spec.ts [hunk 2]
```diff
@@ -293,6 +293,28 @@ describe('modules/manager/npm/extract/index', () => {
       ).toBeArrayIncludingOnly(['https://registry.example.com']);
     });
 
+    it('resolves registry URLs using the package name if set', async () => {
+      fs.readLocalFile.mockImplementation((fileName): Promise<any> => {
```
Comment (secustor) [inline lib/modules/manager/npm/extract/index.spec.ts:297]: Why not simply use `mockResolvedValueOnce`?

Comment (haines) [reply lib/modules/manager/npm/extract/index.spec.ts:297]: That's not equivalent is it? This implementation resolves to `null` for other filenames. It's also consistent with (copied from 😅) the existing tests.

Comment (secustor) [reply lib/modules/manager/npm/extract/index.spec.ts:298]: I wanted to know if you have tried it, because it should work with that too.

**Review by secustor — APPROVED** [review]

**Pull request merged**